### ISSUE TITLE
Training dataset and Architecture datasets seem to be inverted

### ISSUE THREAD

**Issue by Hrayo712** [JaminFong/FNA#17]
Dear @JaminFong 

I hope you're doing well. I have noticed something strange, perhaps its a minor thing, but I would appreciate your input on this.

**on search.py line 109,** you divide the training set (~118K) in two, the training set, which if I understand correctly, you utilize to update the weights of the sampled networks, and the architecture dataset, which you use to update the architecture parameters (alphas):

https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/tools/search.py#L108-L110

These datasets are later used to build dataloaders.

https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/tools/apis/fna_search_apis.py#L91-L99

However, on the training loop within fna_search_runner.py, I see that you access the dataloaders in the reverse order

https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/tools/apis/fna_search_runner.py#L192-L196

This is based on the workflow defined on the config file, which first defines architecture and then training (inverse order as how the dataset and dataloaders are created)

https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/configs/fna_ssdlite_search.py#L186

This ultimately leads to the training set used for updating the architecture parameters, and the architecture set to update the training weights. Is this a bug ? or is just the naming on search.py wrong ?

I was wondering since, after augmentation, the training ends up running with ~59K images on the training set and ~293 on the arhitecture set. (This for SSDLite)

Thanks in advance for your input!

**Comment by JaminFong** [comment]
Hi, thanks for pointing out this. It is just wrong with the naming, which does not affect the training, and I have corrected it. 
> I was wondering since, after agumentation, the training ends up running with 293K images on the training set and 59K on the arhitecture set.

It is right. This is somewhat for the `RepeatDataset` augmentation strategy.
Hoping my answer could help you!

**Comment by Hrayo712** [comment]
Thanks for the clarification @JaminFong ! 

Just one more question:

Why is it then that you are training (updating network weights) with much less data (~59K) than what you use to update the architecture parameters (~293K after augmentation). Couldn't this lead to overfitting ? as you are trying to minimize the architecture loss, with model weights that are updated using much less data. 

I've seen that after search, the training loss is around 4.7, while the architecture loss is still quite high, around 7. Which would suggest this is the case. (I've seen this by printing the loss on every iteration)

I appreciate the support. I am just trying to understand the method better :)

Best!

**Comment by JaminFong** [comment]
Hi, Hrayo!
It is on the contrary. The network weights are updated using more iterations as the config sets. It is common in NAS methods to use more data to update the network weights, as the architecture parameters are fewer than network weights.
For the loss value, the architecture loss not only involves the loss for classification, but also the loss for the cost (e.g. MAdds), so it is larger than the training loss. Moreover, as the super network is much larger than the final searched architecture, the loss during the search is also larger than the final retraining loss.
https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/configs/fna_ssdlite_search.py#L186

**Comment by Hrayo712** [comment]
Hi @JaminFong,

I understand the weights being updated more often ( with more iterations as you show in the config file). However, after the `RepeatDataset`, the `arch_dataset` ends up with ~293K images, which is more data than `train_dataset ` (~59K), meaning you would use more data to update the architecture parameters, than the data you have to update the weights.

https://github.com/JaminFong/FNA/blob/9783a7dd116ecb96c4014030673f98aa027c45fd/fna_det/tools/search.py#L108-L110

Isn't this contradicting to what you mention ?

>  It is common in NAS methods to use more data to update the network weights, as the architecture parameters are fewer than network weights.

I see in the runner that when the batches of a data_loader are "exhausted", the loader is "refilled" (try/except segment). If I understand correctly, this means that within an Epoch (determined by the length of the `arch_dataset`) the `train_dataset ` is actually processed 5 times. Consequently, the `arch_dataset `is processed 5 times less, even though the arch data_loader has more images. Is my understanding correct ? if so, why Repeating the arch dataset in the first place ?

https://github.com/JaminFong/FNA/blob/9783a7dd116ecb96c4014030673f98aa027c45fd/fna_det/tools/apis/fna_search_runner.py#L192-L206

With respect to the loss value, I am only looking at the Cross entropy loss (loss_cls + loss_reg), so I am not taking into account the sub objective loss. But thanks for clarifying that the supernetwork does have an impact on it as well, so its not directly comparable with a given found architecture/child network, right ?

I am also interested in finding out the mAP of the final found network after the search (without retraining). Is there any way to do this directly ?

I tried  introducing some custom code to sample the network, and then process the `train_dataset` (the same used during the search) in batches of 1, to finally use the coco tools to obtain the mAP metrics (I also coded the simple_test function to enable this functionality). However, I am obtaining 0 mAP, even though the loss of the network seems to be quite low by the end of the search (~4 ). Do you know if there any particular reason why this would not work ? or why would the mAP behave like this ?

Thanks in advance for all your input :)

**Comment by JaminFong** [comment]
Hi, Hrayo!
The `RepeatDataset` augmentation is actually lengthening the dataset by repeating it. So I think it may be better to describe it by the iteration numbers instead of data amount. 
> I see in the runner that when the batches of a data_loader are "exhausted", the loader is "refilled" (try/except segment). If I understand correctly, this means that within an Epoch (determined by the length of the arch_dataset) the train_dataset is actually processed 5 times. Consequently, the arch_dataset is processed 5 times less, even though the arch data_loader has more images. 

This understanding is right. The final effect is the network weights are updated 5 times more than architecture parameters.

> I am also interested in finding out the mAP of the final found network after the search. Is there any way to do this directly ?

Please follow the tutorial of "parameter adaptation" in ReadMe.

> However, I am obtaining 0 mAP, even though the loss of the network seems to be quite low by the end of the search (~4 ). Do you know if there any particular reason why this would not work ?

I'm not so clear with the details you do this. Maybe the BN is the point.

**Comment by Hrayo712** [comment]
Hi @JaminFong,

Thanks for all the clarifications.

> I'm not so clear with the details you do this. Maybe the BN is the point.

Indeed this is the case, I believe it is because the supernet BNs are configured to not track the statistics during training, as you can see in the config file below (`affine `and `track `settings configured to False)

https://github.com/JaminFong/FNA/blob/9783a7dd116ecb96c4014030673f98aa027c45fd/fna_det/configs/fna_ssdlite_search.py#L14-L19

Therefore, with this configuration, if at the end of the search, I want to sample the found architecture, and run evaluation on it, it will not give me good results (even though the loss is low), since the batch statistics (running mean, and var) were not recorded during the search process. I did some tests, and believe this to be the reason.

My question is then: Is there an specific reason why this tracking is disabled for the search ? 

I can imagine that since you are sampling several architectures, the statistics will fluctuate, but do you think there would be an issue if I enabled them for the search ? (just to be able to run evaluation on the found architecture at the end of the search, without having to retrain )

I've seen something related to this in the following part of the code. Perhaps you were trying to do something like this ? 

https://github.com/JaminFong/FNA/blob/9783a7dd116ecb96c4014030673f98aa027c45fd/fna_det/models/fna_base_backbone.py#L244-L250

Thanks again for your input :)

**Comment by JaminFong** [comment]
Hi, as the frequent sub-architecture sampling during the search may cause instability to BN updating, BN updating is always disabled in most NAS works. This issue is also discussed in the [FNA++](https://arxiv.org/pdf/2006.12986.pdf) paper. You could refer to Sec. 4.4 and Sec. 5.1 if intersted.

**Comment by Hrayo712** [comment]
Thanks a lot! Closing the issue now :)

Best of luck with your research 👍
